sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -2097,6 +2097,138 @@ def set_application(self, channel, appl_code):
+        for si_keys in optics_si_dict:
+            if si_keys in si_settings:
+                si_settings[si_keys] = optics_si_dict[si_keys]
+
+        tx_si_ctrl_advt_capability = self.get_tx_si_ctrl_advt_capability()
+        if tx_si_ctrl_advt_capability is None:
+            return None
+
+        rx_si_ctrl_advt_capability = self.get_rx_si_ctrl_advt_capability()
+        if rx_si_ctrl_advt_capability is None:
+            return None
+
+        # Apply the SI settings if supported
+        for si_keys in si_settings:
+            if ((si_keys == consts.OUTPUT_EQ_PRE_CURSOR_TARGET_RX and rx_si_ctrl_advt_capability['rx_output_eq_pre_ctrl_supported']) or
+                (si_keys == consts.OUTPUT_EQ_POST_CURSOR_TARGET_RX and rx_si_ctrl_advt_capability['rx_output_eq_post_ctrl_supported']) or
+                (si_keys == consts.OUTPUT_AMPLITUDE_TARGET_RX and rx_si_ctrl_advt_capability['rx_ouput_amp_ctrl_supported']) or
+                (si_keys == consts.FIXED_INPUT_EQ_TARGET_TX and tx_si_ctrl_advt_capability['tx_input_eq_fixed_supported'])):
+                for lane in range(self.NUM_CHANNELS):
+                    if ((1 << lane) & host_lanes_mask) == 0:
+                        continue
+                    lane = lane+1
+                    si_key_lane = "{}{}".format(si_keys, lane)
+                    val = si_settings[si_keys][si_key_lane]
+                    if ((si_keys == consts.OUTPUT_EQ_PRE_CURSOR_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_eq_pre_max_val']) or
+                        (si_keys == consts.OUTPUT_EQ_POST_CURSOR_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_eq_post_max_val']) or
+                        (si_keys == consts.OUTPUT_AMPLITUDE_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_amp_supported_val']) or
+                        (si_keys == consts.FIXED_INPUT_EQ_TARGET_TX and val <= tx_si_ctrl_advt_capability['tx_input_eq_max_val'])) :
+                        if (lane%2) == 0:
+                            pre_si_key_lane = "{}{}".format(si_keys, lane-1)
+                            pre_val = self.xcvr_eeprom.read(pre_si_key_lane)
+                            val = (val << 4) | pre_val
+                        self.xcvr_eeprom.write(si_key_lane, val)
+            elif (si_keys == consts.ADAPTIVE_INPUT_EQ_RECALLED_TX and 
+                 (tx_si_ctrl_advt_capability['tx_input_recall_buf1_supported'] or tx_si_ctrl_advt_capability['tx_input_recall_buf2_supported'])):
+                val = 0
+                for lane in range(self.NUM_CHANNELS):
+                    if ((1 << lane) & host_lanes_mask) == 0:
+                        continue
+                    si_key_lane = "{}{}".format(si_keys, lane+1)
+                    si_val = si_settings[si_keys][si_key_lane]
+                    lane %= (self.NUM_CHANNELS//2)
+                    mask = ~(val << (lane*2))
+                    l_data = si_val << (lane*2)
+                    val = (val & mask) | l_data
```

Comment:
@AnoopKamath its not quite clear, which register we are writing. I would suggest to develop a `nibble_register_write(offset, lane = odd/even, value)` and use that function to write.